### PR TITLE
Total Variation Loss

### LINKED ISSUE #972: Total Variation Loss
## 🚀 Feature

Hi, in the image package, we are missing the [total variation loss](https://en.wikipedia.org/wiki/Total_variation_denoising).

### Motivation

Total variation loss is a useful loss in reconstruction models such as in auto-encoders. Traditionally used in denoising, it is now being used in several other types of hi-res reconstruction.

### PR DESCRIPTION
**Pull request by ragavvenkatesan**
## What does this PR do?

This PR adds Total Variation Loss.
Fixes #972

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@ragavvenkatesan how is it going here? :rabbit: 
we are considering a new feature release next week, do you think we can finish this addition by then so it will be included? :pray:

**Comment by ragavvenkatesan** [comment]
> 

So, I was waiting on you folks for some feedback on the initial code. I am working on tests this weekend, I can prioritize for next release since you brought it up.

**Comment by ragavvenkatesan** [comment]
So, @Borda I think the metric is implemented. I working on unit test (taking suggestions for tests now). I have tested this on actual model training and seems like it is working as expected.

File: torchmetrics/functional/image/total_variation.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        A loss scalar value.
+    """
+
+    def tensor_size(t):
+        return t.size()[1] * t.size()[2] * t.size()[3]
+
+    _height = img.size()[2]
+    _width = img.size()[3]
+    _count_height = tensor_size(img[:, :, 1:, :])
+    _count_width = tensor_size(img[:, :, :, 1:])
```
Comment (SkafteNicki) [inline torchmetrics/functional/image/total_variation.py:21]: this should be equivalent with only one call to the shape function
```suggestion
    _batchsize, _channels, _height, _width = img.shape[1:]
    _count_height = _channels * (_height - 1) * _width
    _count_width = _channels * _height * (_width - 1)
```

File: torchmetrics/functional/image/total_variation.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    _height_tv = torch.pow((img[:, :, 1:, :] - img[:, :, : _height - 1, :]), 2).sum()
+    _width_tv = torch.pow((img[:, :, :, 1:] - img[:, :, :, : _width - 1]), 2).sum()
+    return (2 * (_height_tv / _count_height + _width_tv / _count_width)) / img.size()[0]
```
Comment (SkafteNicki) [inline torchmetrics/functional/image/total_variation.py:24]:
```suggestion
    return (2 * (_height_tv / _count_height + _width_tv / _count_width)) / _batchsize
```

**Review by SkafteNicki — COMMENTED** [review]
some initial comments

File: torchmetrics/functional/image/total_variation.py (hunk 1 above)
Comment (ragavvenkatesan) [reply torchmetrics/functional/image/total_variation.py:21]: Yes, sure. Why `img.shape[1:]` though? Shouldn't it be just `img.shape`? Shape is a 4-element tuple correct?

File: torchmetrics/image/total_variation.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import torch
```
[the remark this answers is not in the document]
Comment (ragavvenkatesan) [reply torchmetrics/image/total_variation.py:15]: Done.

File: torchmetrics/functional/image/total_variation.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import torch
```
[the remark this answers is not in the document]
Comment (ragavvenkatesan) [reply torchmetrics/functional/image/total_variation.py:15]: Done.

File: torchmetrics/image/total_variation.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+from torchmetrics.metric import Metric
+
+
+class TotalVariation(Metric):
+    """Computes Total Variation loss.
```
[the remark this answers is not in the document]
Comment (ragavvenkatesan) [reply torchmetrics/image/total_variation.py:21]: https://en.wikipedia.org/wiki/Total_variation_denoising

**Comment by ragavvenkatesan** [comment]
Addressed all comments @SkafteNicki Anything on tests?

**Comment by SkafteNicki** [comment]
@ragavvenkatesan do we have some kind of reference implementation of this metric (sklearn, scipy, something else)? That is normally what we need for testing to secure that the implementation is correct.

**Comment by ragavvenkatesan** [comment]
> 
I don't think there is an easy implementation available anywhere.
For reference, here is the hacky way that tensorflow tests a similar loss in their pull-request for this loss.

Note: They have slight difference in their implementation as they only use the denoising loss, but I could do something similar.
https://github.com/tensorflow/tensorflow/pull/6662/files#diff-33c59c5ad653d2f5e292da5968484cd7e3e0320dc453f92eb51cc8e1d21ec0d4R2310

**Comment by SkafteNicki** [comment]
Hi @ragavvenkatesan,
It seems like we can compare against:
https://kornia.readthedocs.io/en/latest/_modules/kornia/losses/total_variation.html

**Comment by ragavvenkatesan** [comment]
> 

This is a better idea. I will write tests comparing against this. Thanks.

**Comment by Borda** [comment]
@ragavvenkatesan how is it going? :)

**Comment by ragavvenkatesan** [comment]
Hi @Borda  I apologize, went on vacation. Getting back to this and prioritizing this now.

**Comment by Borda** [comment]
> @Borda can you make a push into the source branch again please? I am getting a merge conflict when I push. Thanks.

all shall be resolved now :)

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@SkafteNicki, any idea why the variable type fails?

**Comment by Borda** [comment]
@ragavvenkatesan, could you pls check and resolve conflicts? :otter:

**Comment by Borda** [comment]
It seems to be failing for all the latest tests; in fact, it hangs and dies...
